File: app/android/app/src/main/java/com/gonezo/multiplatform/audioextraction/contract/ExtractionRequest.java
package com.gonezo.multiplatform.audioextraction.contract;

import java.util.Map;
import org.json.JSONException;
import org.json.JSONObject;

public record ExtractionRequest(
  String schemaVersion,
  Source source,
  Extraction extraction,
  Map<String, Object> context,
  Options options
) {
  public ExtractionRequest {
    if (context == null) {
      context = Map.of();
    }
  }

  public JSONObject toJson() {
    JSONObject result = new JSONObject();
    put(result, "schemaVersion", schemaVersion);
    put(result, "source", source == null ? JSONObject.NULL : source.toJson());
    put(result, "extraction", extraction == null ? JSONObject.NULL : extraction.toJson());
    put(result, "context", ContractJsonMapper.toJsonObject(context));
    if (options != null) {
      put(result, "options", options.toJson());
    }
    return result;
  }

  public static ExtractionRequest fromJson(JSONObject json) {
    if (json == null) {
      return null;
    }
    Source source = Source.fromJson(json.optJSONObject("source"));
    Extraction extraction = Extraction.fromJson(json.optJSONObject("extraction"));
    Options options = Options.fromJson(json.optJSONObject("options"));
    return new ExtractionRequest(
      json.optString("schemaVersion", null),
      source,
      extraction,
      ContractJsonMapper.toMap(json.optJSONObject("context")),
      options
    );
  }

  public record Source(String type, String value) {
    public JSONObject toJson() {
      JSONObject result = new JSONObject();
      put(result, "type", type);
      put(result, "value", value);
      return result;
    }

    public static Source fromJson(JSONObject json) {
      if (json == null) {
        return null;
      }
      return new Source(json.optString("type", null), json.optString("value", null));
    }
  }

  public record Extraction(Map<String, Object> outputSchema, String instructions) {
    public Extraction {
      if (outputSchema == null) {
        outputSchema = Map.of();
      }
    }

    public JSONObject toJson() {
      JSONObject result = new JSONObject();
      put(result, "outputSchema", ContractJsonMapper.toJsonObject(outputSchema));
      if (instructions != null) {
        put(result, "instructions", instructions);
      }
      return result;
    }

    public static Extraction fromJson(JSONObject json) {
      if (json == null) {
        return null;
      }
      return new Extraction(
        ContractJsonMapper.toMap(json.optJSONObject("outputSchema")),
        json.has("instructions") ? json.optString("instructions", null) : null
      );
    }
  }

  public record Options(Boolean includeTranscript, String language) {
    public JSONObject toJson() {
      JSONObject result = new JSONObject();
      if (includeTranscript != null) {
        put(result, "includeTranscript", includeTranscript);
      }
      if (language != null) {
        put(result, "language", language);
      }
      return result;
    }

    public static Options fromJson(JSONObject json) {
      if (json == null) {
        return null;
      }
      return new Options(
        json.has("includeTranscript") ? json.optBoolean("includeTranscript") : null,
        json.has("language") ? json.optString("language", null) : null
      );
    }
  }

  private static void put(JSONObject target, String key, Object value) {
    try {
      target.put(key, value);
    } catch (JSONException ex) {
      throw new IllegalStateException("Cannot serialize extraction request key " + key, ex);
    }
  }
}

File: core/src/main/kotlin/com/gonezo/audioextraction/domain/contract/ExtractionRequest.kt
package com.gonezo.audioextraction.domain.contract

import org.json.JSONObject

data class ExtractionRequest(
  val schemaVersion: String?,
  val source: Source?,
  val extraction: Extraction?,
  val context: Map<String, Any?> = emptyMap(),
  val options: Options? = null,
) {
  fun toJson(): JSONObject {
    val result = JSONObject()
    result.put("schemaVersion", schemaVersion)
    result.put("source", source?.toJson() ?: JSONObject.NULL)
    result.put("extraction", extraction?.toJson() ?: JSONObject.NULL)
    result.put("context", ContractJsonMapper.toJsonObject(context))
    if (options != null) {
      result.put("options", options.toJson())
    }
    return result
  }

  data class Source(
    val type: String?,
    val value: String?,
  ) {
    fun toJson(): JSONObject {
      val result = JSONObject()
      result.put("type", type)
      result.put("value", value)
      return result
    }

    companion object {
      @JvmStatic
      fun fromJson(json: JSONObject?): Source? {
        if (json == null) {
          return null
        }
        return Source(json.optString("type", null), json.optString("value", null))
      }
    }
  }

  data class Extraction(
    val outputSchema: Map<String, Any?> = emptyMap(),
    val instructions: String? = null,
  ) {
    fun toJson(): JSONObject {
      val result = JSONObject()
      result.put("outputSchema", ContractJsonMapper.toJsonObject(outputSchema))
      if (instructions != null) {
        result.put("instructions", instructions)
      }
      return result
    }

    companion object {
      @JvmStatic
      fun fromJson(json: JSONObject?): Extraction? {
        if (json == null) {
          return null
        }
        return Extraction(
          ContractJsonMapper.toMap(json.optJSONObject("outputSchema")),
          if (json.has("instructions")) json.optString("instructions", null) else null,
        )
      }
    }
  }

  data class Options(
    val includeTranscript: Boolean?,
    val language: String?,
  ) {
    fun toJson(): JSONObject {
      val result = JSONObject()
      if (includeTranscript != null) {
        result.put("includeTranscript", includeTranscript)
      }
      if (language != null) {
        result.put("language", language)
      }
      return result
    }

    companion object {
      @JvmStatic
      fun fromJson(json: JSONObject?): Options? {
        if (json == null) {
          return null
        }
        return Options(
          if (json.has("includeTranscript")) json.optBoolean("includeTranscript") else null,
          if (json.has("language")) json.optString("language", null) else null,
        )
      }
    }
  }

  companion object {
    @JvmStatic
    fun fromJson(json: JSONObject?): ExtractionRequest? {
      if (json == null) {
        return null
      }
      return ExtractionRequest(
        json.optString("schemaVersion", null),
        Source.fromJson(json.optJSONObject("source")),
        Extraction.fromJson(json.optJSONObject("extraction")),
        ContractJsonMapper.toMap(json.optJSONObject("context")),
        Options.fromJson(json.optJSONObject("options")),
      )
    }
  }
}
